pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,441 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tcpproxy
+
+import (
+	"errors"
+	"io"
+	"net"
+	"reflect"
+	"runtime/debug"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/timerpool"
+)
+
+// Connection wrap tcp connection
+type Connection struct {
+	rawConn   net.Conn
+	connected uint32
+	closed    uint32
+
+	localAddr  net.Addr
+	remoteAddr net.Addr
+
+	lastBytesSizeRead  int
+	lastWriteSizeWrite int
+
+	readBuffer      []byte
+	writeBuffers    net.Buffers
+	ioBuffers       []*iobufferpool.StreamBuffer
+	writeBufferChan chan *iobufferpool.StreamBuffer
+
+	mu               sync.Mutex
+	startOnce        sync.Once
+	connStopChan     chan struct{} // use for connection close
+	listenerStopChan chan struct{} // use for listener close
+
+	onRead  func(buffer *iobufferpool.StreamBuffer) // execute read filters
+	onClose func(event ConnectionEvent)
+}
+
+// NewDownstreamConn wrap connection create from client
+// @param remoteAddr client addr for udp proxy use
+func NewDownstreamConn(conn net.Conn, remoteAddr net.Addr, listenerStopChan chan struct{}) *Connection {
+	clientConn := &Connection{
+		connected:  1,
+		rawConn:    conn,
+		localAddr:  conn.LocalAddr(),
+		remoteAddr: conn.RemoteAddr(),
+
+		writeBufferChan: make(chan *iobufferpool.StreamBuffer, 8),
+
+		mu:               sync.Mutex{},
+		connStopChan:     make(chan struct{}),
+		listenerStopChan: listenerStopChan,
+	}
+
+	if remoteAddr != nil {
+		clientConn.remoteAddr = remoteAddr
+	} else {
+		clientConn.remoteAddr = conn.RemoteAddr() // udp server rawConn can not get remote address
+	}
+	return clientConn
+}
+
+// LocalAddr get connection local addr
+func (c *Connection) LocalAddr() net.Addr {
+	return c.localAddr
+}
+
+// RemoteAddr get connection remote addr(it's nil for udp server rawConn)
+func (c *Connection) RemoteAddr() net.Addr {
+	return c.rawConn.RemoteAddr()
+}
+
+// SetOnRead set connection read handle
+func (c *Connection) SetOnRead(onRead func(buffer *iobufferpool.StreamBuffer)) {
+	c.onRead = onRead
+}
+
+// OnRead set data read callback
+func (c *Connection) OnRead(buffer *iobufferpool.StreamBuffer) {
+	c.onRead(buffer)
+}
+
+// SetOnClose set close callback
+func (c *Connection) SetOnClose(onclose func(event ConnectionEvent)) {
+	c.onClose = onclose
+}
+
+// Start running connection read/write loop
+func (c *Connection) Start() {
+	c.startOnce.Do(func() {
+		c.startRWLoop()
+	})
+}
+
+// State get connection running state
+func (c *Connection) State() ConnState {
+	if atomic.LoadUint32(&c.closed) == 1 {
+		return ConnClosed
+	}
+	if atomic.LoadUint32(&c.connected) == 1 {
+		return ConnActive
+	}
+	return ConnInit
```

Comment:
propose to use one variable for `c.closed` and `c.connected`, using atomic functions like this is not correct in most cases.